src/internal/storage/fileset/postgres_cache.go

```diff
@@ -0,0 +1,122 @@
+package fileset
+
+import (
+	"context"
+
+	proto "github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+	"github.com/jmoiron/sqlx"
+	"github.com/pachyderm/pachyderm/v2/src/internal/dbutil"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/track"
+)
+
+// CreatePostgresCacheV1 creates the table for a cache.
+// DO NOT MODIFY THIS FUNCTION
+// IT HAS BEEN USED IN A RELEASED MIGRATION
+func CreatePostgresCacheV1(ctx context.Context, tx *pachsql.Tx) error {
+	const schema = `
+	CREATE TABLE storage.cache (
+		key text NOT NULL PRIMARY KEY,
+		value_pb BYTEA NOT NULL,
+		ids UUID[] NOT NULL,
+		accessed_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP
+	);
+`
+	_, err := tx.ExecContext(ctx, schema)
+	return errors.EnsureStack(err)
+}
+
+const CacheTrackerPrefix = "cache/"
+
+func cacheTrackerKey(key string) string {
+	return CacheTrackerPrefix + key
+}
+
+type Cache struct {
+	db      *pachsql.DB
+	tracker track.Tracker
+	maxSize int
+}
+
+func NewCache(db *pachsql.DB, tracker track.Tracker, maxSize int) *Cache {
+	return &Cache{
+		db:      db,
+		tracker: tracker,
+		maxSize: maxSize,
+	}
+}
+
+func (c *Cache) Put(ctx context.Context, key string, value *types.Any, ids []ID) error {
+	data, err := proto.Marshal(value)
+	if err != nil {
+		return errors.EnsureStack(err)
+	}
+	return dbutil.WithTx(ctx, c.db, func(tx *pachsql.Tx) error {
+		if err := c.put(tx, key, data, ids); err != nil {
+			return err
+		}
+		return c.applyEvictionPolicy(tx)
+	})
+}
+
+func (c *Cache) put(tx *pachsql.Tx, key string, value []byte, ids []ID) error {
+	_, err := tx.Exec(`
+		INSERT INTO storage.cache (key, value_pb, ids) 
+		VALUES ($1, $2, $3)
+		ON CONFLICT (key) DO NOTHING
+	`, key, value, ids)
+	if err != nil {
+		return errors.EnsureStack(err)
+	}
+	var pointsTo []string
+	for _, id := range ids {
+		pointsTo = append(pointsTo, id.TrackerID())
+	}
+	return errors.EnsureStack(c.tracker.CreateTx(tx, cacheTrackerKey(key), pointsTo, track.NoTTL))
```

Comment:
Shouldn't the `EnsureStack` be in the tracker, since we also control that code?

Reply:
Unfortunately the wrapcheck linter assumes interfaces could have implementations outside your code, which is part of the reason we have so many. I've been looking into forking it to ignore interfaces we define, but it's a little complicated as it's bundled with `golangci`

Reply:
The linter does a pretty good job ensuring we have a stack at the boundaries of our code and third party libraries, so I am in favor of only using `EnsureStack` when the linter says we need it. Requiring it at our interface boundaries is less than ideal, but it isn't as bad as us just throwing them at arbitrary error propagation sites.

Reply:
Ah okay got it.